### PR TITLE
Allow user to run/debug war in explorer context menu #15

### LINKED ISSUE #15: Should allow "Run on Server" and "Debug on Server" in project/war context menu
Currently we let the user start with the server. You pick a server, choose "Add deployment" etc. It would be a superior user experience if a war or an exploded war in the project tree had a menu entry "Run on Server"/"Debug on Server" to fulfill the same goal.

See what the tomcat adapter for vscode offers ("Debug on tomcat server" in the context menu for a war in the screencast): https://marketplace.visualstudio.com/items?itemName=adashen.vscode-tomcat&ssr=false#overview
image

### PR DESCRIPTION
**Pull request by lstocchi**
it fixes #15 
Test PR

### CODE REVIEW

**Comment by robstryker** [comment]
1)  This patch needs to be rebased. It cannot merge to master
2) Naming any action ".war" is not acceptable, because these actions must support all deployment types. WildFly can publish .war files, .ear files, .rar files, .ejb files, .jar files, .xml files, and many other types as well. So favoring ".war" even in name is not ok.  Also, other runtimes like karaf might support ".kar" files or folders as well.

**Comment by lstocchi** [comment]
I agree with you but the UI doesn't have any knowledge about which file a server can publish so we will end up having a problem here. Do you suggest to just change the name of the action and keep limiting this command to the .war file for now?

**Comment by robstryker** [comment]
No,. because then that makes the action useless for deploying a .jar to an OSGi server.  

It'd probably be more reasonable to allow them to select any file / resource and then any server, and if they get rejected by the server adapter then so be it

**Comment by robstryker** [comment]
To be clear, the reason it should be this way is important to understand. 

A server adapter might not just decide based on extension. That might work for very simply server adapters, but not for all. Some server adapters may need to do things like discover what version of the j2ee servlet spec is there. They may need to look inside the manifest.mf file and look for specific attributes. They might need to scan the directory tree to look for specific files. 

There is almost zero way to properly standardize this over a protocol to let the client do all this before sending it to the server. Sure, I might be able to do something like add a fileNamePattern to the protocol, but that won't cover a lot of use cases. I guess we could then add something like requiredFiles, but that also won't verify every possible scenario. Then if you know a specific file can NOT be present, for example if a server type can not handle deployments with a specific descriptor in it because your server doesn't suppoirt that,  well then we need to add a new pattern. 

Or if we try scanning a web.xml for specific tags that are only valid in the newest version of the spec... well now you'd need to add something like fileMustNotContain file=somePatternName content=someContentPattern.  

And then adding a spec that can work for all of this might just be impossible.

**Comment by lstocchi** [comment]
Modified the old PR. Now there is no restriction from the UI, the commands are always visible so a user can use "Run on Server" action with any file. If an error occured (extension not valid or whatever), the UI will display an error, like the one below returned by the rsp-server.

image

File: package.json [hunk 1]
```diff
@@ -151,6 +151,26 @@
         "command": "server.editServer",
         "title": "Edit Server",
         "category": "Servers"
+      },
+      {
+        "command": "server.distribution.run",
+        "title": "Run on Server",
+        "category": "Distribution"
```
Comment (adietish) [inline package.json]: the naming "Distribution" is confusing to me. Do you mean an application which is bundled as war/ear/jar? To me this is more of an "application artifact(s)", "artifact(s)" or "application".

File: src/serverExplorer.ts [hunk 2]
```diff
@@ -214,51 +215,55 @@ export class ServerExplorer implements TreeDataProvider<RSPState | ServerStateNo
...
+        if (client && file && file.length === 1) {
+
+            const answer = await window.showQuickPick(['No', 'Yes'], {placeHolder:
+                'Do you want to edit optional deployment parameters?'});
+            const options = {};
+            if (!answer) {
+                return;
+            }
+            if (answer === 'Yes') {
+                const deployOptionsResponse:Protocol.ListDeploymentOptionsResponse =
```
Comment (adietish) [inline src/serverExplorer.ts]: I think that I would extract l228-245 to a separate method #getOptions to improve code readability.

**Comment by robstryker** [comment]
Maybe the following:
Run on server (exploded)
Debug on server (exploded)

Or not. The (exploded) might not even be necessary.

**Comment by lstocchi** [comment]
@adietish @robstryker PR updated based on your comments

File: src/extensionApi.ts [hunk 3]
```diff
@@ -543,6 +543,35 @@ export class CommandHandler {
...
+            } else {
+                await this.debugServer(context);
+            }
+        } else if (context.state === ServerState.STARTED) {
+            if (!(context.runMode === ServerState.RUN_MODE_RUN &&
+                mode === ServerState.RUN_MODE_RUN)) {
+                await this.restartServer(mode, context);
+            }
+        } else {
+            return Promise.reject(`Unable to add distribution and run it on server ${context.server.id}. Stop/start the server and try again.`);
```
Comment (adietish) [inline src/extensionApi.ts]: ```distribution```, shouldn't that be ```deployment```?

**Comment by adietish** [comment]
test coverage went down on this. I'd still merge it if we file an issue to add tests?

**Pull request merged**